### PR TITLE
Add support for API Gateway V2 in discovery

### LINKED ISSUE #586: Confirm API Gateway v2 support
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Feature description

From what I can tell, this project currently only supports the REST API Gateway (v1).

Can this be confirmed? In terms of metrics `aws_apigateway_info` is only returning REST instances, although other related metrics do return HTTP gateway values for `dimension_ApiId`.

Looking at the code though, I don't see anything using https://github.com/aws/aws-sdk-go/tree/main/service/apigatewayv2

I would love support for HTTP and WebSocket metrics :)

### What might the configuration look like?

_No response_

### Anything else?

Related #426

### PR DESCRIPTION
**Pull request by matej-g**
Resolves #586.

This PR adds support for metrics generated by API Gateway V2 (https://docs.aws.amazon.com/apigatewayv2/latest/api-reference/api-reference.html) in discovery.

- Adds client and session cache for API Gateway V2 API
- Adds service filter (based on API ID)
- Adds and adjusts dimension regexps for resource association
- Adds counter for API Gateway V2 API requests.

Added tests for service filter, but this has also been tested manually.

### CODE REVIEW

File: pkg/apitagging/filters.go [hunk 1]
```diff
@@ -31,21 +32,34 @@ type serviceFilter struct {
...
+			var (
+				limit           int64 = 500 // max number of results per page. default=25, max=500
+				input                 = apigateway.GetRestApisInput{Limit: &limit}
+				output                = apigateway.GetRestApisOutput{}
+				pageNum         int
+				outputResources []*model.TaggedResource
+			)
+
 			err := client.apiGatewayAPI.GetRestApisPagesWithContext(ctx, &input, func(page *apigateway.GetRestApisOutput, lastPage bool) bool {
+				promutil.APIGatewayAPICounter.Inc()
```
Comment (matej-g) [inline pkg/apitagging/filters.go:46]: I noticed we were incrementing only once per every method call, but I think we want to track the actual number of API calls, so I moved this line to increment on every call.

**Comment by wimsymons** [comment]
Who can re-run the checks so this one can be merged? Seems the linting step failed with a time-out.
I'm waiting on this feature so I can get API Gateway metrics on AWS HTTP API's.

@matej-g can you do that? Or maybe @cristiangreco as he was mentioned earlier.

**Comment by cristiangreco** [comment]
@matej-g can you please rebase? Due to rencent changes to sessioncache and clients (in order to introduce support for sdk v2) you'll need to update your PR again. Sorry for the noise.

File: pkg/config/services.go [hunk 2]
```diff
@@ -92,8 +92,10 @@ var SupportedServices = serviceConfigs{
 			aws.String("apigateway"),
 		},
 		DimensionRegexps: []*regexp.Regexp{
-			regexp.MustCompile("apis/(?P<ApiName>[^/]+)$"),
-			regexp.MustCompile("apis/(?P<ApiName>[^/]+)/stages/(?P<Stage>[^/]+)$"),
+			regexp.MustCompile("restapis/(?P<ApiName>[^/]+)$"),
+			regexp.MustCompile("restapis/(?P<ApiName>[^/]+)/stages/(?P<Stage>[^/]+)$"),
+			regexp.MustCompile("apis/(?P<ApiId>[^/]+)$"),
+			regexp.MustCompile("apis/(?P<ApiId>[^/]+)/stages/(?P<Stage>[^/]+)$"),
```
Comment (cristiangreco) [inline pkg/config/services.go:98]: Can you please add a comment to mark which regexes are for GW v1 vs v2?

**Comment by matej-g** [comment]
@cristiangreco All good, thanks for taking a look! Let me bring this PR up to date and address comments, hopefully we're good to go afterwards.

File: pkg/config/services.go (hunk 2 above)
Comment (matej-g) [reply pkg/config/services.go:115]: Done 👍

File: pkg/clients/tagging/filters_test.go [hunk 3]
```diff
@@ -369,3 +440,12 @@ func (apigateway apiGatewayClient) GetRestApisPagesWithContext(_ aws.Context, _
 	fn(apigateway.getRestApisOutput, true)
 	return nil
 }
+
+type apiGatewayV2Client struct {
+	apigatewayv2iface.ApiGatewayV2API
+	getRestApisOutput *apigatewayv2.GetApisOutput
+}
+
+func (apigateway apiGatewayV2Client) GetApisWithContext(_ aws.Context, input *apigatewayv2.GetApisInput, opts ...request.Option) (*apigatewayv2.GetApisOutput, error) {
```
Comment (wimsymons) [inline pkg/clients/tagging/filters_test.go:449]: @matej-g linter issue remains. It should be fixed here.

Comment (matej-g) [reply pkg/clients/tagging/filters_test.go:449]: I committed wrong change 🤦, fixing it now, thanks!

**Comment by wimsymons** [comment]
Thanks for all the fixes @matej-g ! All checks passed. @cristiangreco are we good to go ?

File: go.mod [hunk 4]
```diff
@@ -24,7 +24,7 @@ require (
 	github.com/jmespath/go-jmespath v0.4.0 // indirect
 	github.com/kr/text v0.2.0 // indirect
 	github.com/matttproud/golang_protobuf_extensions v1.0.4 // indirect
-	github.com/pmezard/go-difflib v1.0.0 // indirect
+	github.com/pmezard/go-difflib v1.0.1-0.20181226105442-5d4384ee4fb2 // indirect
```
Comment (cristiangreco) [inline go.mod:27]: Do you know where this dep update is coming from? If unneeded I'd rather keep it out of the PR please.

File: pkg/config/services.go [hunk 5]
```diff
@@ -107,8 +107,12 @@ var SupportedServices = serviceConfigs{
 			aws.String("apigateway"),
 		},
 		DimensionRegexps: []*regexp.Regexp{
-			regexp.MustCompile("apis/(?P<ApiName>[^/]+)$"),
-			regexp.MustCompile("apis/(?P<ApiName>[^/]+)/stages/(?P<Stage>[^/]+)$"),
+			// DimensionRegexps starting with 'restapis' are for APIGateway V1 gateways (REST API gateways )
```
Comment (cristiangreco) [inline pkg/config/services.go:110]:
```suggestion
			// DimensionRegexps starting with 'restapis' are for APIGateway V1 gateways (REST API gateways)
```

**Comment by cristiangreco** [comment]
LGTM, just a couple of comments.

**Comment by wimsymons** [comment]
Nearly there @matej-g, could you check the comments of @cristiangreco ?

**Review by cristiangreco — APPROVED** [review]
🚀

**Pull request merged**